Rsdev 593 enable fast search

rs-fraser opened the pull request:
This change enables [OpenBabel fast search](https://open-babel.readthedocs.io/en/latest/FileFormats/Fastsearch_format.html) .

New created chemicals get added to 2 files `chemicalsMaster.smi` and `nonIndexedChems.smi`.

There's a batch task setup to run each day at midnight (can be configured as a cron expression), or we can call `/chemistry/index` to update the fast search index. Doing so updates the `fsChemicals.fs` file from `chemicalsMaster.smi`, and deletes the contents of `nonIndexedChems.smi`.

When we perform a search, we use fast search to search `fsChemicals.fs` and "slow" search to search `nonIndexedChems.smi`, which will only contain any chemicals added since the last time indexing ran.

I done some performance testing using this [ChEMBL database](https://chembl.gitbook.io/chembl-interface-documentation/downloads) sdf file, which contains 2.5m chemical elements. 

From the sdf file I created a fast search file, then ran various searches using normal and fast search.

Searching for a simple smiles (which returns lots of results) didn't show a huge improvement. Searching for the smiles "CCC" returned 2million results, and took around 400s for normal search and 300s for fast search. However, a more complex search shows huge performance gains: `c1ccoc1` as the search input returns around 100k results, and takes 390s using "normal" search, and only 13s with fast search.

Commit:
wip: fast search

Commit:
Merge branch 'main' into rsdev-593-enable-fast-search

Commit:
wip: fast search

Commit:
add batch job scheduled update fast search file with any newly added chemicals, and then clear the contents of the non-indexed chemicals file. save incoming chemicals to the master file and non-indexed smiles, so that search can be performed against the fast search file, then the (slower but smaller) non-indexed file. add tests for indexing

Commit:
Merge branch 'main' into rsdev-593-enable-fast-search

Commit:
search correct file. schedule indexing daily at midnight

Commit:
update README

Commit:
add endpoint for triggering indexing. tidy tests. remove limit when using fast search

Commit:
inline index call in tests. formatting.

mKowalski256 commented on src/main/java/com/researchspace/chemistry/search/SearchController.java:
```diff
@@ -37,4 +38,11 @@ public SearchController(SearchService searchService) {
     return searchService.search(searchDTO);
   }
+
+  @GetMapping(value = "/chemistry/index")
```
I think REST-wise that should rather be POST mapping.

rs-fraser replied:
I've updated to POST

mKowalski256 reviewed (commented):
Switch to fast-search index files looks good to me, I can find the chemicals fine.

I've noticed though that PR changes introduce problem with saving my test reaction file `azide alkyne drawings_II.cdx`, which the code in `main` branch has no problems with.

In error.log:
```
2025-04-07T07:38:58.062Z ERROR 1 --- [chemistry] [nio-8090-exec-7] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: com.researchspace.chemistry.ChemistryException: Unable to perform conversion to smiles.] with root cause

com.researchspace.chemistry.ChemistryException: Unable to perform conversion to smiles.
        at com.researchspace.chemistry.convert.ConvertService.lambda$convert$0(ConvertService.java:33) ~[!/:1.0.2]
        at java.base/java.util.Optional.orElseThrow(Unknown Source) ~[na:na]
        at com.researchspace.chemistry.convert.ConvertService.convert(ConvertService.java:30) ~[!/:1.0.2]
        at com.researchspace.chemistry.search.SearchService.getSmilesFromOpenBabel(SearchService.java:175)
[...]
```

... and as an RSpace user I'm seeing:
![image](https://github.com/user-attachments/assets/199ef99f-8eec-46f1-9cd1-a292acfea9c2)

Commit:
don't strip conversion output. use post rather than get for indexing method

rs-fraser commented:
> Switch to fast-search index files looks good to me, I can find the chemicals fine.
> 
> I've noticed though that PR changes introduce problem with saving my test reaction file `azide alkyne drawings_II.cdx`, which the code in `main` branch has no problems with.
> 
> In error.log:
> 
> ```
> 2025-04-07T07:38:58.062Z ERROR 1 --- [chemistry] [nio-8090-exec-7] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: com.researchspace.chemistry.ChemistryException: Unable to perform conversion to smiles.] with root cause
> 
> com.researchspace.chemistry.ChemistryException: Unable to perform conversion to smiles.
>         at com.researchspace.chemistry.convert.ConvertService.lambda$convert$0(ConvertService.java:33) ~[!/:1.0.2]
>         at java.base/java.util.Optional.orElseThrow(Unknown Source) ~[na:na]
>         at com.researchspace.chemistry.convert.ConvertService.convert(ConvertService.java:30) ~[!/:1.0.2]
>         at com.researchspace.chemistry.search.SearchService.getSmilesFromOpenBabel(SearchService.java:175)
> [...]
> ```
> 
> ... and as an RSpace user I'm seeing: ![image](https://private-user-images.githubusercontent.com/8066659/430836631-199ef99f-8eec-46f1-9cd1-a292acfea9c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMjY5MDksIm5iZiI6MTc0NDAyNjYwOSwicGF0aCI6Ii84MDY2NjU5LzQzMDgzNjYzMS0xOTllZjk5Zi04ZWVjLTQ2ZjEtOWNkMS1hMjkyYWNmZWE5YzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDdUMTE1MDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmUxYjc0MWJhYTdlMGVmNTFhZTg4ZmQwNzI2MzU0ZjE4ZTFkMDgxZmI1MGY0ZDk5N2YyYTBkODlhZWM1NDM4YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.AsWVMwYnQKKzbLtB10Hr2PJW9l8tGd77aJbS3xn3rMg)

Should be fixed now. I reverted the call to `String.split()` for the OpenBabel output. It's not immediately obvious why this causes some conversions to fail, but 3 of the `cdx` files I have locally fail when the output is stripped.

mKowalski256 reviewed (approved):
Looks good & works fine for me now.

I've created connected PR on rspace-web side: https://github.com/ResearchSpace-ELN/rspace-web-closed-source/pull/35
